### PR TITLE
Update docker container for CI

### PR DESCRIPTION
**Pull request by zaro0508**
Update circleci docker container with python 3.7 removed and python 3.12 added

depends on https://github.com/Sceptre/sceptre-circleci/pull/19

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -12,5 +12,5 @@ max-complexity = 12
 per-file-ignores =
 	docs/_api/conf.py: E265
 	integration-tests/steps/*: E501,F811,F403,F405
-extend-ignore = E203
+extend-ignore = E201,E202,E203,E231
```
Comment (iainelder) [inline .flake8:15]: [Flake8](https://flake8.pycqa.org/en/latest/user/error-codes.html) told me to check [pycodestyle](https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes) for the meaning of the codes.

* E201: whitespace after `(`
* E202: whitespace before `)`
* E231: missing whitespace after `,`, `;`, or `:`

Why do you add these just now?

File: poetry.lock [hunk 2]
```diff
@@ -1408,24 +1466,24 @@ files = [
...
 
 [extras]
 docs = ["docutils", "sphinx", "sphinx-autodoc-typehints", "sphinx-click", "sphinx-rtd-theme"]
 troposphere = ["troposphere"]
 
 [metadata]
 lock-version = "2.0"
-python-versions = ">=3.8,<3.12"
-content-hash = "04d3d244f177ff9d180c03f7bb0aa066fd8123a90495fec376a89a5b0318a45b"
+python-versions = ">=3.8,<=3.12"
```
Comment (iainelder) [inline poetry.lock:1488]: This is the only change in poetry.lock that I expected.

Why did the a lot of the dependency versions also chgange?

File: sceptre/stack.py [hunk 3]
```diff
@@ -421,7 +421,7 @@ def _set_field_with_deprecated_alias(
 
         if preferred_value and deprecated_value:
             raise InvalidConfigFileError(
-                f"Both '{preferred_config_name}' and '{deprecated_config_name}' are set; You should only set a "
+                f"Both '{preferred_config_name}' and '{deprecated_config_name}' are set, You should only set a "
```
Comment (iainelder) [inline sceptre/stack.py:424]: Is it related to the Docker update, or opportunistic? (If it's opportunistic, I think a period (`.`) works best here.)

File: pyproject.toml [hunk 4]
```diff
@@ -46,7 +46,7 @@ classifiers = [
 "http" = "sceptre.template_handlers.http:Http"
 
 [tool.poetry.dependencies]
-python = ">=3.8,<3.12"
+python = ">=3.8,<=3.12"
```
Comment (iainelder) [inline pyproject.toml:49]: Why extend the range to 3.12 now?

If the intention is to support Python 3.12, then I think we need to make more changes.

Did you see PR https://github.com/Sceptre/sceptre/pull/1390 ? It makes more changes to support Python 3.12.

File: .circleci/config.yml [hunk 5]
```diff
@@ -10,7 +10,7 @@ executors:
 aliases:
   - &docs-job
     docker:
-      - image: sceptreorg/sceptre-circleci:2.0.0
+      - image: sceptreorg/sceptre-circleci:2.1.0
```
Comment (iainelder) [inline .circleci/config.yml:13]: How do you create the new tag for the sceptre-circleci repo?

File: poetry.lock (hunk 2 above)
Comment (zaro0508) [reply poetry.lock:1488]: the only thing i did was run `poetry lock` then `git add poetry.lock`.  i have no idea how poetry works its magic.

**Comment by zaro0508** [comment]
replacing this with PR #1394

**Comment by iainelder** [comment]
If the tests pass, then it looks good to me. I haven't had time to run these locally, but I trust the CI.

The custom image does have all the Python versions we expect.

```console
$ docker run -it --rm sceptreorg/sceptre-circleci:2.1.0 bash -c 'for py in python3.{7,8,9,10,11,12}; do "$py" --version; done'
bash: line 1: python3.7: command not found
Python 3.8.18
Python 3.9.18
Python 3.10.6
Python 3.11.6
Python 3.12.0
```

**Review by iainelder — APPROVED** [review]

**Comment by iainelder** [comment]
I suppose we need to look at #1390  after this to finish adding support for Python 3.12. The main point here is to drop support for Python 3.7.

**Pull request merged**